frame/system/src/migrations/mod.rs

```diff
@@ -0,0 +1,182 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Migrate the reference counting state.
+
+use codec::{Decode, Encode, FullCodec};
+use frame_support::{
+	pallet_prelude::ValueQuery, traits::PalletInfoAccess, weights::Weight, Blake2_128Concat,
+	RuntimeDebug,
+};
+use sp_std::prelude::*;
+
+/// Type used to encode the number of references an account has.
+type RefCount = u32;
+
+/// All balance information for an account.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug, MaxEncodedLen)]
+struct AccountData<Balance> {
+	free: Balance,
+	reserved: Balance,
+	misc_frozen: Balance,
+	fee_frozen: Balance,
+}
+
+/// Information of an account.
+#[derive(Clone, Eq, PartialEq, Default, RuntimeDebug, Encode, Decode)]
+struct AccountInfo<Index, AccountData> {
+	nonce: Index,
+	consumers: RefCount,
+	providers: RefCount,
+	sufficients: RefCount,
+	data: AccountData,
+}
+
+/// Trait to implement to give information about types used for migration
+pub trait V2ToV3 {
+	/// The elections-phragmen pallet.
+	type Pallet: 'static + PalletInfoAccess;
+
+	/// System config account id
+	type AccountId: 'static + FullCodec;
+
+	/// System config index
+	type Index: 'static + FullCodec + Copy;
+
+	/// Currency balance.
+	type Balance: 'static + FullCodec + Copy;
+}
+
+frame_support::generate_storage_alias!(
+	System, UpgradedToU32RefCount => Value<
```

Comment:
I see the problem now... Do you have any suggestion about how to relax this assumption?

Reply:
maybe we can do:
```diff
diff --git a/frame/system/src/migrations/mod.rs b/frame/system/src/migrations/mod.rs
index d1255194b1..d4f4245b94 100644
--- a/frame/system/src/migrations/mod.rs
+++ b/frame/system/src/migrations/mod.rs
@@ -68,12 +68,15 @@ frame_support::generate_storage_alias!(
        >
 );
 
-frame_support::generate_storage_alias!(
-       System, UpgradedToTripleRefCount => Value<
-               bool,
-               ValueQuery
-       >
-);
+struct UpgradedToU32RefCountPrefix<P>;
+impl<P: PalletInfoAccess> frame_support::traits::StorageInstance for UpgradedToU32RefCountPrefix<P> {
+       fn pallet_prefix() -> &'static str {
+               P::name()
+       }
+       const STORAGE_PREFIX: &'static str = "UpgradedToU32RefCount";
+}
+type UpgradedToU32RefCount<P: PalletInfoAccess> = StorageValue<UpgradedToU32RefCountPrefix<P>, bool, ValueQuery>;
+
 
 frame_support::generate_storage_alias!(
        System, Account<T: V2ToV3> => Map<
```

But I think it is fine to use generate_stoarge_alias here because we enforce the name of system pallet. So we can just add a note saying why it is valid (so people don't copy this code in other place where the condition isn't valid).

Reply:
Oh I see: there is a specific trait for this. I will remember this since it will probably be useful later.
Though I kept the name and added comments, since I have not yet dug into the trait definition.